device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
     this.on('removeListener', (eventName, eventCallback) => {
-      if (eventName === 'message') {
-        if (this._c2dLink) {
-          if (this.listeners('message').length === 0) {
-            this._c2dLink.detach((err) => {
-              if (err) {
-                debug('error detaching c2d link: ' + err.toString());
-              } else {
-                this._c2dLink.removeListener('error', c2dErrorListener);
-                this._c2dLink.removeListener('message', c2dMessageListener);
-                this._c2dLink = undefined;
-              }
-            });
+      if (eventName === 'message' && this.listeners('message').length === 0) {
+        debug('automatically detaching the link after the last message listener was removed');
+        this._fsm.handle('detachC2DLink', (err) => {
+          if (err) {
+            this.emit('errorReceived', err);
           }
-        } else {
-          debug('No active C2D link: nothing to do here.');
-        }
+        });
+      } else {
+        debug('removing listener for event: ' + eventName);
       }
     });
 
     /*Codes_SRS_NODE_DEVICE_AMQP_RECEIVER_16_003: [The `Amqp` object shall listen to the `message` and error events of the underlying `ReceiverLink` object when it has listeners on its `message` event.]*/
     this.on('newListener', (eventName, eventCallback) => {
       debug('registering new event handler for: ' + eventName);
       if (eventName === 'message') {
-        if (!this._c2dLink) {
-          this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, link) => {
-            this._c2dLink = link;
-            this._c2dLink.on('error', c2dErrorListener);
-            this._c2dLink.on('message', c2dMessageListener);
+        if (this.listeners('message').length === 0) {
+          debug('automatically attaching the link after the first message listener was registered');
+          this._fsm.handle('attachC2DLink', (err) => {
+            debug('C2D link attached automatically');
+            if (err) {
+              debug('error attaching the C2D link automatically: ' + err.toString());
+              this.emit('errorReceived', err);
+            }
           });
-        } else {
-          debug('c2d link already active: nothing to do here.');
         }
+      } else {
+        debug('adding listener for event: ' + eventName);
+      }
+    });
+
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err, result) => {
+            if (callback) {
+              if (err) {
+                callback(err);
+              } else {
+                callback(null, result);
+              }
+            }
+          },
+          connect: (connectCallback) => this._fsm.transition('connecting', connectCallback),
+          disconnect: (disconnectCallback) => {
+            if (disconnectCallback) {
+              disconnectCallback(null, new results.Disconnected());
```

Comment:
>results.Disconnected() [](start = 43, length = 22)

Does this mean "I was already disconnected" or "I am now disconnected".  If the second, why doesn't _onEnter also pass this?

Reply:
you are correct that _onEnter should use results.Disconnected().

---
In reply to: [141127383](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141127383) [](ancestors = 141127383)